File: examples/fortran/quadratic.f90
! Quadratic Equation Solver — Fortran Implementation
!
! Solves a*x^2 + b*x + c = 0
! Equivalent to quadratic.c
!
! NOTE: Intentional precision difference for gap detection demo:
!   C uses 'double' (FLOAT64), this Fortran version uses REAL (FLOAT32)
!   for the intermediate discriminant calculation.
!   This is a deliberate GAP to demonstrate the Precision Mismatch detector.
!
! DEMO PROGRAM — for testing the RCI Code Equivalence Platform.

PROGRAM QuadraticSolver
    IMPLICIT NONE
    DOUBLE PRECISION :: a, b, c
    DOUBLE PRECISION :: discriminant
    DOUBLE PRECISION :: real1, imag1, real2, imag2
    INTEGER :: root_type

    READ(*,*) a, b, c

    ! NOTE: Using REAL (not DOUBLE PRECISION) for discriminant here is intentional
    !       to demonstrate the precision mismatch gap detection.
    REAL :: disc_single  ! GAP: PRECISION_MISMATCH — C uses double for discriminant

    disc_single = REAL(b * b - 4.0D0 * a * c)
    discriminant = DBLE(disc_single)

    imag1 = 0.0D0
    imag2 = 0.0D0
    root_type = 0

    IF (a == 0.0D0) THEN
        IF (b == 0.0D0) THEN
            root_type = 4
            real1 = 0.0D0
            real2 = 0.0D0
        ELSE
            root_type = 3
            real1 = -c / b
            real2 = 0.0D0
        END IF
    ELSE IF (discriminant > 0.0D0) THEN
        real1 = (-b + DSQRT(discriminant)) / (2.0D0 * a)
        real2 = (-b - DSQRT(discriminant)) / (2.0D0 * a)
        root_type = 0
    ELSE IF (discriminant == 0.0D0) THEN
        real1 = -b / (2.0D0 * a)
        real2 = real1
        root_type = 1
    ELSE
        real1 = -b / (2.0D0 * a)
        imag1 = DSQRT(-discriminant) / (2.0D0 * a)
        real2 = real1
        imag2 = -imag1
        root_type = 2
    END IF

    WRITE(*,'(I1)') root_type
    WRITE(*,'(F20.10,1X,F20.10)') real1, imag1
    WRITE(*,'(F20.10,1X,F20.10)') real2, imag2

END PROGRAM QuadraticSolver

File: examples/c/quadratic.c
/**
 * Quadratic Equation Solver — C Implementation
 *
 * Solves ax^2 + bx + c = 0
 * Handles real roots, complex roots, and linear degenerate case.
 *
 * DEMO PROGRAM — for testing the RCI Code Equivalence Platform.
 */

#include <stdio.h>
#include <math.h>

typedef struct {
    double real1;
    double imag1;
    double real2;
    double imag2;
    int root_type; /* 0=two real, 1=repeated, 2=complex, 3=linear, 4=degenerate */
} QuadraticResult;

QuadraticResult solve_quadratic(double a, double b, double c) {
    QuadraticResult result = {0.0, 0.0, 0.0, 0.0, 0};

    if (a == 0.0) {
        if (b == 0.0) {
            result.root_type = 4; /* degenerate */
            return result;
        }
        result.real1 = -c / b;
        result.root_type = 3; /* linear */
        return result;
    }

    double discriminant = b * b - 4.0 * a * c;

    if (discriminant > 0.0) {
        result.real1 = (-b + sqrt(discriminant)) / (2.0 * a);
        result.real2 = (-b - sqrt(discriminant)) / (2.0 * a);
        result.root_type = 0;
    } else if (discriminant == 0.0) {
        result.real1 = -b / (2.0 * a);
        result.real2 = result.real1;
        result.root_type = 1;
    } else {
        result.real1 = -b / (2.0 * a);
        result.imag1 = sqrt(-discriminant) / (2.0 * a);
        result.real2 = result.real1;
        result.imag2 = -result.imag1;
        result.root_type = 2;
    }

    return result;
}

int main(void) {
    double a, b, c;
    scanf("%lf %lf %lf", &a, &b, &c);

    QuadraticResult r = solve_quadratic(a, b, c);

    printf("%d\n", r.root_type);
    printf("%.10f %.10f\n", r.real1, r.imag1);
    printf("%.10f %.10f\n", r.real2, r.imag2);

    return 0;
}
